restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java

```diff
@@ -390,6 +391,122 @@ private Object convertToBackendBooleanValue(boolean value, boolean numericBoolea
+
+    db = maybeCreateTableAndIndexes(dbFactory, db, keyspace, collection, headers, authToken);
+    List<String> idsWritten = new ArrayList<>();
+    try (BufferedReader reader = new BufferedReader(new InputStreamReader(payload, "UTF-8"))) {
+      Iterator<String> iter = reader.lines().iterator();
+      ExecutionContext context = ExecutionContext.NOOP_CONTEXT;
+      String docsPath = convertToJsonPtr(idPath.get());
+      int chunkIndex = 0;
+      final int CHUNK_SIZE = 250;
+      while (iter.hasNext()) {
+        chunkIndex++;
+        Map<String, String> docsInChunk = new LinkedHashMap<>();
+        while (docsInChunk.size() < CHUNK_SIZE && iter.hasNext()) {
+          String doc = iter.next();
+          JsonNode json = mapper.readTree(doc);
+
+          String docId = UUID.randomUUID().toString();
+          if (idPath.isPresent()) {
+            if (!json.at(docsPath).isTextual()) {
+              throw new ErrorCodeRuntimeException(
+                  ErrorCode.DOCS_API_WRITE_BATCH_INVALID_ID_PATH,
+                  String.format(
+                      "Json Document %s requires a String value at the path %s, found %s."
+                          + " Batch %d failed, %d writes were successful. Repeated requests are "
+                          + "idempotent if the same `idPath` is defined.",
+                      doc,
+                      idPath.get(),
+                      json.at(docsPath).toString(),
+                      chunkIndex,
+                      (chunkIndex - 1) * CHUNK_SIZE));
+            }
+            docId = json.requiredAt(docsPath).asText();
+          }
+          docsInChunk.put(docId, doc);
+        }
+
+        // Write the chunk of (at most) 250 documents by firing a single batch with the row inserts
+        List<Object[]> bindVariableList = new ArrayList<>();
+        DocumentDB finalDb = db;
+        List<String> ids =
+            docsInChunk.entrySet().stream()
+                .map(
+                    data -> {
+                      bindVariableList.addAll(
+                          shredJson(
+                                  surfer,
+                                  finalDb,
+                                  Collections.emptyList(),
+                                  data.getKey(),
+                                  data.getValue(),
+                                  false)
+                              .left);
+                      return data.getKey();
+                    })
+                .collect(Collectors.toList());
+
+        long now = timeSource.currentTimeMicros();
+        try {
+          db.deleteManyThenInsertBatch(
+              keyspace,
+              collection,
+              ids,
+              bindVariableList,
+              Collections.emptyList(),
+              now,
+              context.nested("ASYNC INSERT"));
+        } catch (Exception e) {
+          throw new ErrorCodeRuntimeException(
+              ErrorCode.DOCS_API_WRITE_BATCH_FAILED,
```

Comment:
We do not wait for the `CompletableFuture` associated with the batch operation (or did I miss that), so this catch will not reflect storage-level failure. Would it make sense to wait for batch acknowledgement so that a successful HTTP response would mean the batch was accepted (we should probably use the `202` response code in that case)? WDYT?

Reply:
It seems like `deleteManyThenInsertBatch` calls `executeBatch`, which then calls

``` 
if (useLoggedBatches) {
      dataStore.batch(queries, ConsistencyLevel.LOCAL_QUORUM).join();
} else {
      dataStore.unloggedBatch(queries, ConsistencyLevel.LOCAL_QUORUM).join();
}
```

This would mean that it is already blocking on some sort of ack, right? If the ack may not mean that data was written, then I agree with changing this to 202...but most probably that should be changed in the other write endpoints too, which I would consider out of scope for this PR; so I think I will leave as 201 just for now, to remain consistent with the other document write endpoints

Reply:
Ah, right, I missed that :facepalm: